### PR TITLE
[Perf] Extend Qwen Triton warmup to avoid first-request latency spikes

### PR DESCRIPTION
**Pull request by vhagor**
# [Perf] Extend Qwen Triton warmup to avoid first-request latency spikes

## Purpose

Extend `qwen_triton_warmup` so remaining Qwen3.5 / Qwen3-Next Triton kernels compile **before** the JIT monitor is armed, instead of on the first live request. That first-request compile spike is a large hit for latency-sensitive serving.

The warnings below were captured on **pooling** Qwen3.5 (embedding / classification). **Generate** VL hits most of the same kernels (`_bilinear_pos_embed_kernel`, `rotary_kernel`, `layer_norm_fwd_kernel`, plus `triton_mrope` / `batch_memcpy_kernel`); dummy runs do not enumerate those compile keys either. The extra pooling-only gap is GDN prefill (`_causal_conv1d_fwd_kernel`, `_fused_post_conv_kernel`), which generate warmup on main already covers but pooling skipped via `if runner.is_pooling_model: return`.

This continues #40137, #36599, #37338, #47539, #47546 and is complementary to #48363. It addresses leftover JIT-monitor warnings from #43009 (`_bilinear_pos_embed_kernel`, `rotary_kernel`, `_causal_conv1d_fwd_kernel`, `_fused_post_conv_kernel`, `layer_norm_fwd_kernel`, plus `batch_memcpy_kernel` / M-RoPE).

Fixes part of #43009.

## Problem

Observed on **pooling-mode** Qwen3.5 (embedding / classification). After engine warmup, the first live pooling request paid Triton JIT; the second reused the in-process cache and was **>2s faster** on L20. A multi-second first-request spike is a large hit for latency-sensitive serving (online embedding / classification, tight SLO, first-token). `jit_monitor` (#40137) logged:

```
WARNING 08-26 04:52:46 [jit_monitor.py:103] Triton kernel JIT compilation during inference: _bilinear_pos_embed_kernel. This causes a latency spike; consider extending warmup to cover this shape/config.
WARNING 08-26 04:52:46 [jit_monitor.py:103] Triton kernel JIT compilation during inference: rotary_kernel. This causes a latency spike; consider extending warmup to cover this shape/config.
WARNING 08-26 04:52:47 [jit_monitor.py:103] Triton kernel JIT compilation during inference: _causal_conv1d_fwd_kernel. This causes a latency spike; consider extending warmup to cover this shape/config.
WARNING 08-26 04:52:47 [jit_monitor.py:103] Triton kernel JIT compilation during inference: _fused_post_conv_kernel. This causes a latency spike; consider extending warmup to cover this shape/config.
WARNING 08-26 04:52:48 [jit_monitor.py:103] Triton kernel JIT compilation during inference: layer_norm_fwd_kernel. This causes a latency spike; consider extending warmup to cover this shape/config.
```

**Generate** Qwen3.5-VL uses most of these same kernels on the first real request. Dummy max-token / CUDA-graph runs do not compile `_bilinear_pos_embed_kernel`, `rotary_kernel`, `layer_norm_fwd_kernel`, `triton_mrope`, or `batch_memcpy_kernel` either. The new vision / M-RoPE / RMSNormGated / memcpy launchers therefore apply to **both** generate and pooling.

What pooling uniquely exposed: `qwen_triton_warmup` on main already warms GDN causal-conv, fused post-conv (`L ∈ {1, 2, 16}`), and the decode sigmoid-gating kernel (#47539 / #47546) for **generate**, then `if runner.is_pooling_model: return` skipped the whole function. That is why `_causal_conv1d_fwd_kernel` and `_fused_post_conv_kernel` still JIT on the first pooling request even though generate warmup already covers them.

1. **Pooling was skipped entirely** (`if runner.is_pooling_model: return`). See [Pooling](#pooling).
2. **Vision interpolate + rotary.** `_bilinear_pos_embed_kernel` / `rotary_kernel` specialize on grid H/W 16-divisibility (generate VL and pooling VL). Related: #47637 (`do_not_specialize` on H/W); this PR warms the kernels instead of changing specialization.
3. **M-RoPE.** CUDA hits `triton_mrope` only with 2-D positions `(3, T)`. A 1-D dummy falls through to `apply_rotary_emb` and never compiles the live kernel. `T ∈ {1, 2, 16}` covers Triton’s `==1` / `%16==0` / other integer buckets. Reads `uses_mrope` / head counts from the runner or `model_config` so V1 and V2 both work.
4. **RMSNormGated** (`layer_norm_fwd_kernel`). GDN calls it after `reshape(-1, head_v_dim)`, so production `M = num_tokens * hv`. Triton’s key is `(M specialization, ROWS_PER_BLOCK)` plus gated `HAS_Z=True`. Dummy max-batch only covers one bucket. Scanning `1..max_num_tokens` would also cover the keys but is an O(N) startup tax; warmup should enumerate the `{1, %16==0, other} × ROWS_PER_BLOCK` classes instead.
5. **`batch_memcpy_kernel`.** Only launched on the Mamba/GDN prefix-cache state copy (`do_mamba_copy_block`). Dummy / CUDA graph / GDN post-conv never reach it. Triton’s key includes pointer dtypes (`src=*u64, dst=*u64, sizes=*i32`). Warmup must use `int32` sizes or a different cubin is compiled and the first prefix-cache hit still JITs.

## Test plan

```bash
pytest tests/model_executor/test_qwen_triton_warmup.py -q
```

Serve Qwen3.5 9B on L20 with `--jit-monitor-mode=warn` (cold Triton cache). The production dump was **pooling**; also check **generate** VL, which uses most of the same kernels.

Compare first vs second request latency. Confirm no `Triton kernel JIT compilation during inference` for:

- `_bilinear_pos_embed_kernel` (generate + pooling VL)
- `rotary_kernel` (generate + pooling VL)
- `layer_norm_fwd_kernel` (generate + pooling)
- `_causal_conv1d_fwd_kernel` / `_fused_post_conv_kernel` (pooling; generate already warmed these)
- `batch_memcpy_kernel` / `triton_mrope`

## Test result

- Unit tests: `.venv/bin/python -m pytest tests/model_executor/test_qwen_triton_warmup.py -v` — 13 passed, including optional-`norm` config, bounded RMSNorm key enumeration, and skip-RMSNorm-when-no-norm.
- Pooling-mode Qwen3.5 9B on L20: first-request extra delay vs the second request dropped from **>2s**.
- Those pooling-path jit_monitor warnings are gone. Generate VL shares the same new launchers for most of them:
  - `_bilinear_pos_embed_kernel` (generate + pooling)
  - `rotary_kernel` (generate + pooling)
  - `layer_norm_fwd_kernel` (generate + pooling)
  - `_causal_conv1d_fwd_kernel` / `_fused_post_conv_kernel` (pooling; was skipped by the early return)
  - `batch_memcpy_kernel` (int32 `sizes` so the prefix-cache cubin matches) and `triton_mrope`
- Generate/pooling VL : no >2s first-packet spike.

### CODE REVIEW

File: vllm/model_executor/warmup/qwen_triton_warmup.py [hunk 1]
```diff
@@ -241,20 +398,137 @@ def _warm_fused_sigmoid_gating_delta_rule_update_kernel(
...
+    logger.info("Warmed Mamba/GDN batch_memcpy_kernel.")
+
+
 def _synchronize_device(device: torch.device) -> None:
     if device.type == "cuda":
         torch.accelerator.synchronize(device)
 
 
+def _warm_vision(model: torch.nn.Module) -> None:
+    from vllm.model_executor.models.qwen3_vl import Qwen3_VisionTransformer
```
Comment (Isotr0py) [inline vllm/model_executor/warmup/qwen_triton_warmup.py]: I don't think we should warmup ViT and MRoPE here, this file is for Qwen GDN warmup. Please uses another file for ViT warmup instaed and include Qwen3-VL.

File: vllm/model_executor/warmup/qwen_triton_warmup.py [hunk 2]
```diff
@@ -241,20 +398,137 @@ def _warm_fused_sigmoid_gating_delta_rule_update_kernel(
     )
 
 
+def _warm_batch_memcpy_kernel(device: torch.device) -> None:
+    """Warm the Mamba/GDN prefix-cache state copy specialization."""
+    if device.type != "cuda":
+        return
```
Comment (Isotr0py) [inline vllm/model_executor/warmup/qwen_triton_warmup.py]: Ditto, I think all Mamba-style models should warmup this.

File: vllm/model_executor/warmup/qwen_triton_warmup.py (hunk 1 above)
Comment (vhagor) [reply vllm/model_executor/warmup/qwen_triton_warmup.py]: OK, I moved ViT + M-RoPE to `qwen_vl_triton_warmup.py`. It is no longer gated on the GDN `model_type` set; it probes for `Qwen3_VisionTransformer` / `MRotaryEmbedding`, so Qwen3-VL, Qwen3-VL-MoE, and Qwen3.5-VL all share the same path.

File: vllm/model_executor/warmup/qwen_triton_warmup.py (hunk 2 above)
Comment (vhagor) [reply vllm/model_executor/warmup/qwen_triton_warmup.py]: OK, I moved it to `mamba_triton_warmup.py` and run it for any runner with a `MambaSpec` group.

**Comment by vhagor** [comment]
@Isotr0py Thanks for the feedback — I’ve updated the code accordingly.

File: vllm/model_executor/warmup/qwen_vl_triton_warmup.py [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+    model_config = getattr(runner, "model_config", None)
+    parallel_config = getattr(runner, "parallel_config", None)
+    return int(model_config.get_num_attention_heads(parallel_config))
+
+
+def _warm_mrope(runner: "GPUModelRunner", model: torch.nn.Module) -> None:
+    if not _runner_uses_mrope(runner):
+        return
+
+    from vllm.model_executor.layers.rotary_embedding.mrope import MRotaryEmbedding
```
Comment (Isotr0py) [inline vllm/model_executor/warmup/qwen_vl_triton_warmup.py]: I think we should warmup MRoPE for all models with `SupportsMRoPE`?

File: vllm/model_executor/warmup/qwen_triton_warmup.py [hunk 4]
```diff
@@ -66,7 +66,6 @@ def _is_qwen_gdn_layer(module: object) -> bool:
             "conv_kernel_size",
             "tp_size",
             "kv_cache",
-            "norm",
```
Comment (Isotr0py) [inline vllm/model_executor/warmup/qwen_triton_warmup.py:69]: Why?

File: vllm/model_executor/warmup/qwen_vl_triton_warmup.py (hunk 3 above)
Comment (Isotr0py) [inline vllm/model_executor/warmup/qwen_vl_triton_warmup.py:64]: Should we warm mrope for all models with `SupportMRoPE`?

File: vllm/model_executor/warmup/qwen_vl_triton_warmup.py [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+    return bool(getattr(model_config, "uses_mrope", False))
+
+
+def _runner_num_query_heads(runner: object) -> int:
+    num_query_heads = getattr(runner, "num_query_heads", None)
+    if num_query_heads is not None:
+        return int(num_query_heads)
+    model_config = getattr(runner, "model_config", None)
+    parallel_config = getattr(runner, "parallel_config", None)
+    return int(model_config.get_num_attention_heads(parallel_config))
```
Comment (Isotr0py) [inline vllm/model_executor/warmup/qwen_vl_triton_warmup.py]: Is it really necessary to implement such helper functions with so many edge cases `getattr` fallback?

File: vllm/model_executor/warmup/qwen_vl_triton_warmup.py [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+            rope.head_size,
+            rope.rotary_dim,
+            tuple(rope.mrope_section or ()),
+            rope.mrope_interleaved,
+            rope.is_neox_style,
+            runner.dtype,
+        )
+        if key in seen:
+            continue
+        seen.add(key)
```
Comment (Isotr0py) [inline vllm/model_executor/warmup/qwen_vl_triton_warmup.py]: Is it a valid concern that a model has multiple MRoPE configs?

File: vllm/model_executor/warmup/qwen_triton_warmup.py (hunk 4 above)
Comment (vhagor) [reply vllm/model_executor/warmup/qwen_triton_warmup.py:69]: I found Qwen GDN (QwenGatedDeltaNetAttention) always builds RMSNormGated in __init__, so keeping norm in the probe would just tighten the match. I already restored that check.

File: vllm/model_executor/warmup/qwen_vl_triton_warmup.py (hunk 3 above)
Comment (vhagor) [reply vllm/model_executor/warmup/qwen_vl_triton_warmup.py:64]: Yes. triton_mrope is shared, so M-RoPE warmup is now gated on supports_mrope(model) instead of Qwen3-VL only. Vision warmup is still Qwen3_VisionTransformer only.

File: vllm/model_executor/warmup/qwen_vl_triton_warmup.py (hunk 5 above)
Comment (vhagor) [reply vllm/model_executor/warmup/qwen_vl_triton_warmup.py]: Those helpers had many getattr fallbacks. I already removed them.

File: vllm/model_executor/warmup/qwen_vl_triton_warmup.py (hunk 6 above)
Comment (vhagor) [reply vllm/model_executor/warmup/qwen_vl_triton_warmup.py]: Same as above. I dropped the seen set and now warm the first module only.

**Review by Isotr0py — APPROVED** [review]
